draco/analysis/dayenu.py

```diff
@@ -0,0 +1,746 @@
+        * np.cos(m_center * dra)
+    )
+
+    uflag, uindex = np.unique(flag.reshape(-1, nra), return_inverse=True, axis=0)
+    uflag = uflag[:, np.newaxis, :] & uflag[:, :, np.newaxis]
+    uflag = uflag.astype(np.float64)
+
+    ucov = uflag * cov[np.newaxis, :, :]
+
+    pinv = np.linalg.pinv(ucov, hermitian=True) * uflag
+    pinv = pinv[uindex, :, :].reshape(oshp)
+
+    return pinv
+
+
+def lowpass_mmode_filter(ra, m_cut, flag, epsilon=1e-10):
+    """Construct a low-pass m-mode filter.
+
+    The pass band will range from [-m_cut, m_cut].
+
+    Parameters
+    ----------
+    ra : np.ndarray[nra,]
+        Righ ascension in radians.
+    m_cut : float
+        The half width of the pass band.
+    flag : np.ndarray[nfreq, nra]
+        Boolean flag that indicates what right ascensions are valid
+        as a function of frequency.
+    epsilon : float
+        The stop-band rejection of the filter.  Defaults to 1e-10.
+
+    Returns
+    -------
+    pinv : np.ndarray[nfreq, nra, nra]
+        Low-pass m-mode filter for each frequency.
+    """
+    ishp = flag.shape
+    nra = ra.size
+    assert ishp[-1] == nra
+
+    oshp = ishp + (nra,)
+
+    a = np.median(np.abs(np.diff(ra))) * m_cut / np.pi
+    aeps = a * epsilon
+
+    dra = ra[:, np.newaxis] - ra[np.newaxis, :]
+
+    cov = np.eye(nra, dtype=np.float64) / aeps
+    cov += a * (1.0 - 1.0 / aeps) * np.sinc(m_cut * dra / np.pi)
+
+    uflag, uindex = np.unique(flag.reshape(-1, nra), return_inverse=True, axis=0)
+    uflag = uflag[:, np.newaxis, :] & uflag[:, :, np.newaxis]
+    uflag = uflag.astype(np.float64)
+
+    ucov = uflag * cov[np.newaxis, :, :]
+
+    pinv = np.linalg.pinv(ucov, hermitian=True) * uflag
+    pinv = pinv[uindex, :, :].reshape(oshp)
+
+    return pinv
+
+
+def highpass_mmode_filter(ra, m_cut, flag, epsilon=1e-10):
+    """Construct a high-pass m-mode filter.
+
+    The stop band will range from [-m_cut, m_cut].
+
+    Parameters
+    ----------
+    ra : np.ndarray[nra,]
+        Righ ascension in radians.
+    m_cut : float
+        The half width of the stop band.
+    flag : np.ndarray[nfreq, nra]
+        Boolean flag that indicates what right ascensions are valid
+        as a function of frequency.
+    epsilon : float
+        The stop-band rejection of the filter.  Defaults to 1e-10.
+
+    Returns
+    -------
+    pinv : np.ndarray[nfreq, nra, nra]
+        High-pass m-mode filter for each frequency.
+    """
+    ishp = flag.shape
+    nra = ra.size
+    assert ishp[-1] == nra
+
+    oshp = ishp + (nra,)
+
+    dra = ra[:, np.newaxis] - ra[np.newaxis, :]
+
+    cov = np.eye(nra, dtype=np.float64)
+    cov += np.sinc(m_cut * dra / np.pi) / epsilon
+
+    uflag, uindex = np.unique(flag.reshape(-1, nra), return_inverse=True, axis=0)
+    uflag = uflag[:, np.newaxis, :] & uflag[:, :, np.newaxis]
+    uflag = uflag.astype(np.float64)
+
+    ucov = uflag * cov[np.newaxis, :, :]
+
+    pinv = np.linalg.pinv(ucov, hermitian=True) * uflag
```

Comment:
You're probably better using the `scipy.linalg` version as it'll me MKL accelerated.

Reply:
I changed this but then realized that the reason I used the numpy version in the first place is because it accepts a stack of matrices, whereas the scipy version does not.  We stack up all the unique sets of flags and construct the filter with one call, whereas with the scipy version we would have to loop over the unique sets of flags.  I'm not sure how much that speeds up things (if at all) compared to the gains you get from the MKL accelerated version.  In our current scenario we only have one set of flags since complete sidereal coverage is required for the m-mode transform, so scipy with loop is probably faster.  I'm working on timing this for different scenarios now.

Reply:
That's annoying.

Usually the numpy versions are slower implementations and single threaded, so you'd expect ~8x speed up if the matrices are "large enough" (probably anything over N>200 is big enough). However, it might just not be worth the time to fix it at the moment as we're going to refactor everything at somepoint. Could just put a comment about why the numpy versions are being used and the leave it at that.

Reply:
I was able to get the total run time of the task down from 35 minutes to 2 minutes (using 4 nodes).

It turned out that applying the filter to the map was very slow.  The filter was being applied to a view of the map in the container at a specific (beam, pol, el) that was not contiguous in memory.  Creating a contiguous array first resulted in a factor of 200 speed up in the filter application.

In the case where we have a single set of frequency flags for all right ascensions, the filter construction is only 5-10% of the total run time (after the fix above).  But if we ever use more complicated masks that change as a function of RA, then that will grow linearly with the number of unique sets of flags.

I went ahead and switched to using the scipy version of pinv.  However my testing actually suggested that this is 5-10% slower than the numpy version.  As mentioned above, this doesn't really matter for our current usage since filter construction is a small fraction of total run time.  I don't want to spend much more time on this now, but we may want to investigate scipy versus numpy a bit more when we refactor this code.

Reply:
Ah nice. That sounds awesome!